# Stage 1


# Accelerating Cleantech Advancements through NLP-Powered Text Mining


## Data Collection and Cleaning Tutorial for Cleantech NLP Project


In this tutorial, we will go through the steps of collecting and cleaning data for our Cleantech NLP project. We have two main datasets: the Cleantech Media Dataset and the Cleantech Google Patent Dataset.


## Setting up the Environment


We begin by mounting the Google Drive to access the datasets stored there. This step is necessary to read the data into our Colab notebook for analysis and cleaning.


## Loading the Data


Load the datasets into pandas DataFrames to prepare for exploration and cleaning.

The media dataset is in CSV format, while the patent dataset is in JSON format, reflecting the structured nature of patent data.


In [71]:
# Samuels drive link

import pandas as pd

# Paths to the datasets
cleantech_media_path = "cleantech_media_dataset_v2_2024-02-23.csv"
cleantech_patent_path = "cleantech_rag_evaluation_data_2024-02-23.csv"

In [72]:
# Loading the datasets
cleantech_media_data = pd.read_csv(cleantech_media_path)
cleantech_patent_data = pd.read_csv(cleantech_patent_path)

## Initial Data Exploration

Before cleaning, we need to understand our data, which includes checking the shape, examining the columns, and identifying any immediate issues such as missing values or inconsistent data types.


In [73]:
# Display the initial shape of the datasets
print("Initial Cleantech Media Data shape:", cleantech_media_data.shape)
print("Initial Cleantech Patent Data shape:", cleantech_patent_data.shape)

Initial Cleantech Media Data shape: (9593, 8)
Initial Cleantech Patent Data shape: (23, 5)


In [74]:
# Display the first few rows of the media dataset
cleantech_media_data.head()


Unnamed: 0                                              title        date  \
0        1280  Qatar to Slash Emissions as LNG Expansion Adva...  2021-01-13   
1        1281               India Launches Its First 700 MW PHWR  2021-01-15   
2        1283              New Chapter for US-China Energy Trade  2021-01-20   
3        1284  Japan: Slow Restarts Cast Doubt on 2030 Energy...  2021-01-22   
4        1285     NYC Pension Funds to Divest Fossil Fuel Shares  2021-01-25   

  author                                            content       domain  \
0    NaN  ["Qatar Petroleum ( QP) is targeting aggressiv...  energyintel   
1    NaN  ["• Nuclear Power Corp. of India Ltd. ( NPCIL)...  energyintel   
2    NaN  ["New US President Joe Biden took office this ...  energyintel   
3    NaN  ["The slow pace of Japanese reactor restarts c...  energyintel   
4    NaN  ["Two of New York City's largest pension funds...  energyintel   

                                                 url Unnamed: 7  
0  https://www.energyintel.com/0000017b-a7dc-de4c...        NaN  
1  https://www.energyintel.com/0000017b-a7dc-de4c...        NaN  
2  https://www.energyintel.com/0000017b-a7dc-de4c...        NaN  
3  https://www.energyintel.com/0000017b-a7dc-de4c...        NaN  
4  https://www.energyintel.com/0000017b-a7dc-de4c...        NaN

In [75]:
# Display the column names
print(cleantech_media_data.columns, "\n")

# Check the content of the first row
print(cleantech_media_data["content"][0], "\n")

# Check the type of the content column
cleantech_media_data["content"].apply(type).value_counts()


Index(['Unnamed: 0', 'title', 'date', 'author', 'content', 'domain', 'url',
       'Unnamed: 7'],
      dtype='object') 

["Qatar Petroleum ( QP) is targeting aggressive cuts in its greenhouse gas emissions as it prepares to launch Phase 2 of its planned 48 million ton per year LNG expansion. In its latest Sustainability Report published on Wednesday, QP said its goals include `` reducing the emissions intensity of Qatar's LNG facilities by 25% and of its upstream facilities by at least 15%. '' The company is also aiming to reduce gas flaring intensity across its upstream facilities by more than 75% and has raised its carbon capture and storage ambitions from 5 million tons/yr to 7 million tons/yr by 2027. About 2.2 million tons/yr of the carbon capture goal will come from the 32 million ton/yr Phase 1 of the LNG expansion, also known as the North Field East project. A further 1.1 million tons/yr will come from Phase 2, known as the North Field South project, which will raise Qatar's L

content
<class 'str'>      9551
<class 'float'>      42
Name: count, dtype: int64

In [76]:
# Filter out the float values
print(
    cleantech_media_data["content"][
        cleantech_media_data["content"].apply(type) == float
    ]
)

# Print the number of float values
len(
    cleantech_media_data["content"][
        cleantech_media_data["content"].apply(type) == float
    ]
)

2874    NaN
2943    NaN
3832    NaN
4907    NaN
4911    NaN
4916    NaN
4972    NaN
4975    NaN
4996    NaN
5011    NaN
5047    NaN
5048    NaN
5062    NaN
5100    NaN
5145    NaN
5175    NaN
5191    NaN
5194    NaN
5201    NaN
5203    NaN
5240    NaN
5253    NaN
5256    NaN
5285    NaN
5303    NaN
5313    NaN
5396    NaN
5400    NaN
5403    NaN
5408    NaN
5441    NaN
5447    NaN
5458    NaN
5481    NaN
5517    NaN
5520    NaN
5539    NaN
5558    NaN
6716    NaN
6724    NaN
6730    NaN
6741    NaN
Name: content, dtype: object


42

In [77]:
# Print missing values in each column
print("Title: ", cleantech_media_data["title"].isna().sum())

print("Date: ", cleantech_media_data["date"].isna().sum())

print("Author: ", cleantech_media_data["author"].isna().sum())

print("Content: ", cleantech_media_data["content"].isna().sum())

print("Domain: ", cleantech_media_data["domain"].isna().sum())

print("Url: , ", cleantech_media_data["url"].isna().sum())


Title:  0
Date:  0
Author:  9520
Content:  42
Domain:  0
Url: ,  0


## Inspect the Data Structure


In [78]:
# Get a concise summary of the DataFrame
print(cleantech_patent_data.info())

# Display statistical summary for numerical columns
print(cleantech_patent_data.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   example_id      23 non-null     int64 
 1   question_id     23 non-null     int64 
 2   question        23 non-null     object
 3   relevant_chunk  23 non-null     object
 4   article_url     23 non-null     object
dtypes: int64(2), object(3)
memory usage: 1.0+ KB
None
       example_id  question_id
count    23.00000    23.000000
mean     12.00000    10.869565
std       6.78233     6.348259
min       1.00000     1.000000
25%       6.50000     5.500000
50%      12.00000    11.000000
75%      17.50000    16.500000
max      23.00000    21.000000


## Data Cleaning


### Remove Duplicates

Identify and remove any duplicate entries in the datasets to prevent biased analyses.


In [79]:
# Check for duplicates in the media data
print("Duplicates in media data:", cleantech_media_data.duplicated().sum())

Duplicates in media data: 0


Remove duplicates based on 'publication_number' as it should uniquely identify each patent
This ensures that each patent is represented only once in the dataset


In [80]:
# Checking and removing duplicates in the patent data based on 'publication_number'
duplicate_counts = cleantech_patent_data.duplicated(
    subset="publication_number").sum()
print(
    f"Number of duplicate rows based on publication_number: {duplicate_counts}")

KeyError: Index(['publication_number'], dtype='object')

In [ ]:
# Remove these duplicate rows
cleantech_patent_data = cleantech_patent_data.drop_duplicates(
    subset="publication_number"
)

The patent dataset contains columns with list-like structures, making them unhashable and problematic for drop_duplicates().
We need to convert these columns to a hashable type (like string):


In [ ]:
# Further deduplication in patent data based on text content

# Convert list-type columns to string for the purpose of identifying duplicates
cleantech_patent_data["abstract_localized_str"] = cleantech_patent_data[
    "abstract_localized"
].apply(str)
print(
    "Duplicates in patent data based on text content:",
    cleantech_patent_data.duplicated(subset=["abstract_localized_str"]).sum(),
)

# Check for duplicates in the patent dataset
patent_duplicates = cleantech_patent_data.duplicated(subset=["abstract_localized_str"])
print("Number of duplicate entries in patent data:", patent_duplicates.sum())

Duplicates in patent data based on text content: 87
Number of duplicate entries in patent data: 87


In [ ]:
# Drop duplicates
cleantech_patent_data = cleantech_patent_data.drop_duplicates(
    subset=["abstract_localized_str"]
)

# Cleaning up by dropping the temporary string column
cleantech_patent_data = cleantech_patent_data.drop(columns=["abstract_localized_str"])

# Verify the operation
print("Final shape of patent data after cleaning:", cleantech_patent_data.shape)

Final shape of patent data after cleaning: (13325, 8)


## Handle Nested JSON Fields


We have the fields title_localized and abstract_localized are nested, so extract relevant information:


In [ ]:
# Extracting 'text' from 'title_localized' and 'abstract_localized'
cleantech_patent_data["title"] = cleantech_patent_data["title_localized"].apply(
    lambda x: x[0]["text"] if x else None
)
cleantech_patent_data["abstract"] = cleantech_patent_data["abstract_localized"].apply(
    lambda x: x[0]["text"] if x else None
)

# Check the result
print(cleantech_patent_data[["title", "abstract"]].head())

                                               title  \
0  Adaptable DC-AC Inverter Drive System and Oper...   
1  System for providing the energy from a single ...   
2      Verfahren zum steuern einer windenergieanlage   
4  Control method for optimizing solar-to-power e...   
5  Low-carbon running saline wastewater treatment...   

                                            abstract  
0  Disclosed is an adaptable DC-AC inverter syste...  
1  In accordance with an example embodiment, a so...  
2  Verfahren zum Steuern einer Windenergieanlage ...  
4  A control method for optimizing a solar-to-pow...  
5  本发明公开了一种应用太阳能和环路热管的低碳运行含盐废水处理系统及方法，属于含盐废水低碳处理领...  


## Handling Missing Values

Determine if any columns have a significant number of missing values and decide whether to impute or remove these entries.


In [ ]:
# Check for missing values before removal
print("Missing values in media data:")
print(cleantech_media_data.isnull().sum())

Missing values in media data:
Unnamed: 0       0
title            0
date             0
author        9562
content          0
domain           0
url              0
dtype: int64


In [ ]:
# Dropping the 'Unnamed: 0' column as it appears to be an auto-generated index with no intrinsic meaning.
cleantech_media_data = cleantech_media_data.drop(columns=["Unnamed: 0"])

This column likely originated from an index in the source file and does not provide useful information for analysis.


In [ ]:
# Dropping the 'author' column if not needed
cleantech_media_data.drop("author", axis=1, inplace=True)

Dropping the 'author' column due to a large number of missing values (9562 out of total).
Given that the authorship is not central to our analysis of cleantech trends and innovations, 
and imputation of these missing values is not feasible without reliable auxiliary information, 
we remove this column to focus on more complete and relevant data aspects.


In [ ]:
# Final verification of data structure and types
print(cleantech_media_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9593 entries, 0 to 9592
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    9593 non-null   object
 1   date     9593 non-null   object
 2   content  9593 non-null   object
 3   domain   9593 non-null   object
 4   url      9593 non-null   object
dtypes: object(5)
memory usage: 374.9+ KB
None


In [ ]:
# Show the count of missing values in the patent dataset
print("Missing values in patent dataset:")
print(cleantech_patent_data.isnull().sum())

Missing values in patent dataset:
publication_number    0
application_number    0
country_code          0
title_localized       0
abstract_localized    0
publication_date      0
inventor              0
cpc                   0
title                 0
abstract              0
dtype: int64


In [ ]:
# Replace empty lists or None with NaN
cleantech_patent_data = cleantech_patent_data.applymap(lambda x: pd.NA if not x else x)

# Check for missing values
print(cleantech_patent_data.isnull().sum())

publication_number       0
application_number       0
country_code             0
title_localized          0
abstract_localized       0
publication_date         0
inventor              4973
cpc                   4867
title                    0
abstract                 0
dtype: int64


## Data Type Consistency

Ensure that all columns are of the correct data type, such as converting dates to datetime objects for easier analysis later


In [ ]:
# Convert date columns to datetime format
cleantech_media_data["date"] = pd.to_datetime(cleantech_media_data["date"])

# Verify the data types
print(cleantech_media_data.dtypes)

title              object
date       datetime64[ns]
content            object
domain             object
url                object
dtype: object


In [ ]:
# Convert 'publication_date' to datetime format
cleantech_patent_data["publication_date"] = pd.to_datetime(
    cleantech_patent_data["publication_date"], errors="coerce"
)

# Verify the conversion
print(cleantech_patent_data["publication_date"].head())

0   1970-01-01 00:00:00.020220728
1   1970-01-01 00:00:00.020220728
2   1970-01-01 00:00:00.020220727
4   1970-01-01 00:00:00.020220726
5   1970-01-01 00:00:00.020220722
Name: publication_date, dtype: datetime64[ns]


## Final Data Overview

After cleaning, we provide an overview of the cleaned datasets to verify that they are ready for analysis.


In [ ]:
print(cleantech_media_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9593 entries, 0 to 9592
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype         
---  ------   --------------  -----         
 0   title    9593 non-null   object        
 1   date     9593 non-null   datetime64[ns]
 2   content  9593 non-null   object        
 3   domain   9593 non-null   object        
 4   url      9593 non-null   object        
dtypes: datetime64[ns](1), object(4)
memory usage: 374.9+ KB
None


In [ ]:
print(cleantech_patent_data.info())

<class 'pandas.core.frame.DataFrame'>
Index: 13325 entries, 0 to 29997
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   publication_number  13325 non-null  object        
 1   application_number  13325 non-null  object        
 2   country_code        13325 non-null  object        
 3   title_localized     13325 non-null  object        
 4   abstract_localized  13325 non-null  object        
 5   publication_date    13325 non-null  datetime64[ns]
 6   inventor            8352 non-null   object        
 7   cpc                 8458 non-null   object        
 8   title               13325 non-null  object        
 9   abstract            13325 non-null  object        
dtypes: datetime64[ns](1), object(9)
memory usage: 1.1+ MB
None


## Saving the Cleaned Datasets


In [ ]:
# After completing all data cleaning steps

# Specify the path for saving the cleaned datasets
cleaned_media_data_path = "/content/drive/My Drive/CLT/cleantech_media_cleaned.csv"
cleaned_patent_data_path = "/content/drive/My Drive/CLT/cleantech_patent_cleaned.csv"

# Save the cleaned media dataset
cleantech_media_data.to_csv(cleaned_media_data_path, index=False)
print(f"Cleaned media data saved to {cleaned_media_data_path}")

# Save the cleaned patent dataset
cleantech_patent_data.to_csv(cleaned_patent_data_path, index=False)
print(f"Cleaned patent data saved to {cleaned_patent_data_path}")

Cleaned media data saved to /content/drive/My Drive/CLT/cleantech_media_cleaned.csv
Cleaned patent data saved to /content/drive/My Drive/CLT/cleantech_patent_cleaned.csv


# Text Preprocessing


## Importing Necessary Libraries

We start by importing the libraries needed for data manipulation and text preprocessing.


In [ ]:
# Import necessary libraries
import pandas as pd
import nltk
import re
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

In [ ]:
# Ensure that NLTK's tokenizers and stopwords data are available
nltk.download("punkt")
nltk.download("stopwords")
nltk.download("wordnet")

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


True

### Text Preprocessing Function

This function will be used to clean and preprocess text data.


In [ ]:
def preprocess_text(text):
    """
    Function to preprocess text data by lowering case, removing punctuation,
    tokenizing, removing stopwords, and lemmatizing.
    """
    # Convert to lowercase
    text = text.lower()
    # Remove punctuation and special characters
    text = re.sub(r"[^\w\s]", "", text)
    # Tokenization
    tokens = nltk.word_tokenize(text)
    # Remove stopwords
    stop_words = set(stopwords.words("english"))
    tokens = [word for word in tokens if word not in stop_words]
    # Lemmatization
    lemmatizer = WordNetLemmatizer()
    tokens = [lemmatizer.lemmatize(word) for word in tokens]
    return " ".join(tokens)

### Loading and Preprocessing the Text


In [ ]:
# Define the paths where the cleaned datasets were saved
media_data_cleaned_path = "/content/drive/My Drive/CLT/cleantech_media_cleaned.csv"
patent_data_cleaned_path = "/content/drive/My Drive/CLT/cleantech_patent_cleaned.csv"

# Loading the cleaned media dataset
media_data = pd.read_csv(media_data_cleaned_path)
print("Cleaned Cleantech Media Data loaded successfully.")
print(media_data.head())  # Display the first few rows to verify the data

# Loading the cleaned patent dataset
patent_data = pd.read_csv(patent_data_cleaned_path)
print("\nCleaned Cleantech Patent Data loaded successfully.")
print(patent_data.head())  # Display the first few rows to verify the data

Cleaned Cleantech Media Data loaded successfully.
                                               title        date  \
0  Qatar to Slash Emissions as LNG Expansion Adva...  2021-01-13   
1               India Launches Its First 700 MW PHWR  2021-01-15   
2              New Chapter for US-China Energy Trade  2021-01-20   
3  Japan: Slow Restarts Cast Doubt on 2030 Energy...  2021-01-22   
4     NYC Pension Funds to Divest Fossil Fuel Shares  2021-01-25   

                                             content       domain  \
0  ["Qatar Petroleum ( QP) is targeting aggressiv...  energyintel   
1  ["• Nuclear Power Corp. of India Ltd. ( NPCIL)...  energyintel   
2  ["New US President Joe Biden took office this ...  energyintel   
3  ["The slow pace of Japanese reactor restarts c...  energyintel   
4  ["Two of New York City's largest pension funds...  energyintel   

                                                 url  
0  https://www.energyintel.com/0000017b-a7dc-de4c...  
1  https://www.e

### Applying Text Preprocessing

We now apply the defined text preprocessing function to relevant columns in the datasets.


In [ ]:
# Preprocess 'title' and 'content' columns in the media dataset
media_data["title_preprocessed"] = media_data["title"].apply(preprocess_text)
media_data["content_preprocessed"] = media_data["content"].apply(preprocess_text)

In [ ]:
# Displaying a sample of the preprocessed data
print("Media Data Preprocessed Sample:")
print(
    media_data[
        ["title", "title_preprocessed", "content", "content_preprocessed"]
    ].head()
)

Media Data Preprocessed Sample:
                                               title  \
0  Qatar to Slash Emissions as LNG Expansion Adva...   
1               India Launches Its First 700 MW PHWR   
2              New Chapter for US-China Energy Trade   
3  Japan: Slow Restarts Cast Doubt on 2030 Energy...   
4     NYC Pension Funds to Divest Fossil Fuel Shares   

                                title_preprocessed  \
0       qatar slash emission lng expansion advance   
1                   india launch first 700 mw phwr   
2                 new chapter uschina energy trade   
3  japan slow restarts cast doubt 2030 energy plan   
4        nyc pension fund divest fossil fuel share   

                                             content  \
0  ["Qatar Petroleum ( QP) is targeting aggressiv...   
1  ["• Nuclear Power Corp. of India Ltd. ( NPCIL)...   
2  ["New US President Joe Biden took office this ...   
3  ["The slow pace of Japanese reactor restarts c...   
4  ["Two of New York City'

In [ ]:
# Preprocessing the 'title_localized' and 'abstract_localized' columns in the patent dataset
patent_data["title_preprocessed"] = patent_data["title_localized"].apply(
    preprocess_text
)
patent_data["abstract_preprocessed"] = patent_data["abstract_localized"].apply(
    preprocess_text
)

# Display a sample of the preprocessed patent data to verify
print("\nPatent Data Preprocessed Sample:")
print(patent_data[["title_preprocessed", "abstract_preprocessed"]].head())


Patent Data Preprocessed Sample:
                                  title_preprocessed  \
0  text adaptable dcac inverter drive system oper...   
1  text system providing energy single contiguous...   
2  text verfahren zum steuern einer windenergiean...   
3  text control method optimizing solartopower ef...   
4  text lowcarbon running saline wastewater treat...   

                               abstract_preprocessed  
0  text disclosed adaptable dcac inverter system ...  
1  text accordance example embodiment solar energ...  
2  text verfahren zum steuern einer windenergiean...  
3  text control method optimizing solartopower ef...  
4  text 本发明公开了一种应用太阳能和环路热管的低碳运行含盐废水处理系统及方法属于含盐废水低...  


## Saving Text preprocessed Data


In [ ]:
# Define the paths for saving the preprocessed datasets
preprocessed_media_path = "/content/drive/My Drive/CLT/cleantech_media_preprocessed.csv"
preprocessed_patent_path = (
    "/content/drive/My Drive/CLT/cleantech_patent_preprocessed.csv"
)

# Save the preprocessed media dataset to a new CSV file
media_data.to_csv(preprocessed_media_path, index=False)
print(f"Preprocessed media dataset saved to: {preprocessed_media_path}")

# Save the preprocessed patent dataset to a new CSV file
patent_data.to_csv(preprocessed_patent_path, index=False)
print(f"Preprocessed patent dataset saved to: {preprocessed_patent_path}")

Preprocessed media dataset saved to: /content/drive/My Drive/CLT/cleantech_media_preprocessed.csv
Preprocessed patent dataset saved to: /content/drive/My Drive/CLT/cleantech_patent_preprocessed.csv


# Conclusion

The text data from both the Cleantech Media Dataset and the Cleantech Google Patent Dataset has been successfully preprocessed.
